Add FormattedText property support to Label in Controls

Redth opened the pull request:
Brings support for Label.FormattedText in Controls.

Todo:
 - [x] Remove compatibility extensions and use Controls'
 - [x] Tests

Redth committed:
Initial Label.FormattedText work

Redth committed:
Update formattedstring / font / control extensions

Redth committed:
FormattedText label sample

Redth committed:
Check for NaN as a default font size too

Redth committed:
Remove somewhat duplicate code

Redth committed:
Remove CompatServiceProvider

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Search harder for service

Redth committed:
Add more mocks, fix tests

Redth committed:
Merge branch 'main' into label-formattedtext

jsuarezruiz commented:
/azp run

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Stop using Device.GetNamedSize for default Font sizes in some places

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Also check if Attributed string is empty, not just null

Redth committed:
Whoops, check the virtual control's value not the platform

Redth committed:
TextTransform should default to label's value

If a span has no transform specified (default), it should defer to the value set on the label itself.

This didn't work this way in Forms as far as I can tell but should really be the case.

Redth committed:
Added some text transform tests for formattedstring

Redth committed:
Fix windows default transform

Redth committed:
Use platform specific test property in tests to retrieve text

Redth committed:
Cleanup platform extensions

Redth committed:
Add a test for FormattedString platform conversion

Redth committed:
Tests asserting default font size should consult the platform defaults

Redth committed:
Fix more assumptions that default font size is always 10 in tests

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Use FindMauiContext

Redth committed:
Ensure we have a handler to get a fontmanager with

Redth committed:
Merge branch 'main' into label-formattedtext

mattleibow commented on src/Compatibility/Core/src/Forms.cs:
```diff
@@ -11,6 +11,8 @@ public static void Init(IActivationState activationState)
 			throw new NotImplementedException();
 		}
+
+		public static IMauiContext MauiContext => throw new NotImplementedException();
```
Should this be internal? I see it is internal in the platforms.

mattleibow reviewed (commented):
Thigs feel OK. My main question is about the DefaultSize property on the FontManager. There is no way to override and also the shared implementation is -1.

Is it not possible to have this be an instance? Already for Windows it was.

PureWeen committed:
- cleanup shell ctor

PureWeen commented on src/Compatibility/Core/src/Windows/FontImageSourceHandler.cs:
```diff
@@ -17,6 +18,10 @@ public sealed class FontImageSourceHandler : IImageSourceHandler, IIconElementHa
 
+		IFontManager _fontManager;
+		IFontManager FontManager
+			=> _fontManager ??= Application.Current?.FindMauiContext()?.Services?.GetService<IFontManager>();
```
You could use `imagesource` as the starting point to search for the MauiContext 

`imagesource.FindMauiContext`

PureWeen commented on src/Compatibility/Core/src/Android/Renderers/FontImageSourceHandler.cs:
```diff
@@ -5,13 +5,18 @@
 	public sealed class FontImageSourceHandler : IImageSourceHandler
 	{
+		IFontManager _fontManager;
+		IFontManager FontManager => _fontManager ??= Application.Current?.FindMauiContext()?.Services?.GetService<IFontManager>();
```
You could use `imagesource` as the starting point to search for the MauiContext 

`imagesource.FindMauiContext`

PureWeen committed:
- simplify shell ctors

Redth committed:
Make font size an instance member of FontManager

This also creates a RequireFontManager view extension to help get an instance of font manager.

There's a parameter which allows this call to fall back to using Application.Current in cases where there's no MAUI Context found (ie: No handler attached yet when ApplyPropertiesVisitor runs).

Redth committed:
Move property to internal

Redth commented on src/Controls/samples/Controls.Sample/Pages/Controls/LabelPage.xaml:
```diff
@@ -45,6 +45,33 @@
+                    <FormattedString>
+                        <Span Text="Plain old Text" />
+                        <Span Text=" " />
+                        <Span Text="Colors" BackgroundColor="Cyan" TextColor="Navy" />
```
Yeah I put a comment in the code, it seems the WinUI API doesn't support a background color in a `Run` https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.documents.run?view=winui-3.0#properties

Redth commented on src/Core/src/Fonts/FontManager.Standard.cs:
```diff
@@ -5,6 +5,8 @@ namespace Microsoft.Maui
 {
 	public class FontManager : IFontManager
 	{
+		public static double DefaultFontSize => -1;
```
Yep, this is the right call... I've reworked things so that this is an instance getter on the font manager.

Redth commented on src/Controls/tests/Core.UnitTests/MockPlatformServices.cs:
```diff
@@ -34,7 +34,7 @@ public double GetNamedSize(NamedSize size, Type targetElement, bool useOldSizes)
 			{
 				case NamedSize.Default:
-					return 10;
+					return Maui.FontManager.DefaultFontSize;
```
This will be 10 again now in tests, as the MockFontManager now returns 10 after moving this back to be an instance property.

Redth committed:
Use a GetDefaultFontSize extension

This will return 0 (just as FontManager was already doing) when it cannot locate a font manager service.

The reason this needs to happen is because the Default Values Creator runs for some controls for FontSize before there's ever a Maui Context available.

Redth committed:
Default value of 0 to match the DVC fallback

Redth committed:
Update tests

Redth committed:
Remove incompatible packages

Redth committed:
Make compat calls accept fontregistrar instance

Since MAUI Context will not be created and registered yet, but we can new up a font loader and font registrar since it's the compat gallery that passes it in anyway.

Redth committed:
Remove old overridden target that interferes with remote sim for iOS

Redth committed:
Merge branch 'main' into label-formattedtext

Redth committed:
Fix up xaml unit tests

Redth committed:
Fix more tests

Redth committed:
Merge branch 'main' into label-formattedtext

softlion commented:
Why not add a converter from list<span> to formattedstring, in order make the formatted string tag optional ?
